sql/2019/10_SEO/10_05.sql

```diff
@@ -0,0 +1,13 @@
+#standardSQL
+
+#ld+json, microformatting, schema.org + what @type
+
+SELECT
+    url,
+    flattened_105
+FROM
+    `httparchive.pages.2019_07_01_desktop`
+CROSS JOIN
+    UNNEST(CAST(JSON_EXTRACT_SCALAR(REPLACE(JSON_EXTRACT_SCALAR(payload, '$._almanac'), '\\"','"'), "$['10.5']") as ARRAY)) as flattened_105
```

Comment:
@rviscomi For this query I need to count the occurrences in our dataset of each element in the array _almanac['10.5']. I try to unnest (and then GROUP BY and COUNT).

But the CAST of the JSON_EXTRACT_SCALAR fails somewhere, and I don't know why.

Reply:
Here's an example using a smaller 1k example table:

```sql
CREATE TEMPORARY FUNCTION parseStructuredData(payload STRING)
RETURNS ARRAY<STRING> LANGUAGE js AS '''
  try {
    var $ = JSON.parse(payload);
    var almanac = JSON.parse($._almanac);
    return Array.from(almanac['10.5']);
  } catch (e) {
    return [];
  }
''';

SELECT
  url,
  parseStructuredData(payload)
FROM
  `httparchive.almanac.pages_desktop_1k`
```

I'm using a User Defined Function here to do the parsing and array conversion. BigQuery and JSON arrays don't play well together.

Reply:
Perfect. Makes sense.